src/embed/http.rs

```diff
@@ -0,0 +1,91 @@
+use std::convert::TryFrom;
+use std::error::Error;
+use std::fmt;
+use std::str;
+use wasm_bindgen::JsCast;
+use wasm_bindgen::JsValue;
+use wasm_bindgen_futures::JsFuture;
+use web_sys::{RequestInit, RequestMode};
+
+// browser_fetch makes and HTTP request over the network via the browser's Fetch API.
+// It is intended to be used in wasm; it won't work in regular rust. The response contains
+// the status and body, it doesn't populate response headers (though it easily could).
+// Since we can't run a web server in wasm land and this fn doesn't work in rust land
+// this fn is tested manually in tests/wasm.rs.
+//
+// TODO timeout/abort
+// TODO log request/response
+pub async fn browser_fetch(
+    http_req: &http::Request<String>,
+) -> Result<http::Response<String>, FetchError> {
+    let mut opts = RequestInit::new();
+    opts.method(http_req.method().as_str());
+    opts.mode(RequestMode::Cors);
+    let js_body = JsValue::from_str(http_req.body());
+    opts.body(Some(&js_body));
+    let web_sys_req = web_sys::Request::new_with_str_and_init(&http_req.uri().to_string(), &opts)?;
+    let h = web_sys_req.headers();
+    for (k, v) in http_req.headers().iter() {
+        h.set(k.as_ref(), v.to_str().unwrap())?;
```

Comment:
```bash
$ su phritz
$ echo "BUT THE PAAAANICS"
```

Reply:
(why not just make this one of the `FetchError`s?)

Reply:
panics indeed, nice catch. either i missed this on my sweep of cleaups or i had a type assertion thingy above, dunno.

Reply:
PS - I looked into this and I think the problem is somewhat lessened in Rust/web. Panics take down the wasm vm, but not the web page. So the webpage will still "work" modulo replicache (so probably not -- heh). You could also theoretically reload the wasm in that case.

OTOH, given the enumerated errors, I really feel a lot less sympathy for having any panics in the codebase at all, as I feel it is easy enough to handle them correctly.